explainer.md

```diff
@@ -1,624 +1,397 @@
-</head>
-<body>
-  <div class="animated" sharedid="header-id" id="header">Shared Element</div>
-</body>
-</html>
+<img alt="Page-A and Page-B" src="media/pages.png?raw=true">
+
+The transition between Page-A and Page-B can be a full cross-document navigation between two same-origin pages, or it can be an SPA navigation. The main mechanism and concepts are the same between the two.
+
+Cross-origin transitions are something we want to tackle, but may have significant differences and restrictions for security reasons. Cross-origin transitions are not covered in this document.
+
+## Part 1: The offering
+
+Before Page-A goes away, it offers up parts of itself to be used in the transition. Generally, this will mean one part per 'thing' that acts independently during the transition. For the example transition, the parts are:
+
+- The header
+- The share button
+- The rest (referred to as the page root)
+
+https://user-images.githubusercontent.com/93594/141104275-6d1fb67a-2f73-41e4-9cef-14676798223b.mp4
+
+Aside from the root, an element can only be offered as a transition part if it has paint containment, which clips child content to the content box, although this can be expanded with `overflow-clip-margin`. The is a compromise made to greatly simplify implementation.
+
+Additionally, the element must be a single rect. As in, it doesn't break across lines or columns.
+
+- Open question: If the developer tries to break the above rules, what should happen? Is paint containment applied by the browser? Is that element dropped from the transition? Is the whole transition abandoned?
+
+When a developer offers part of the page for transitions, there are two modes they can choose from:
+
+### As a single texture
+
+The entire painting of the element is captured, including things which appear outside of its bounding box such as shadows and blurs, as a single texture at device-pixel resolution.
+
+https://user-images.githubusercontent.com/93594/141118353-d62d19a1-0964-4fa0-880f-bdde656ce899.mp4
+
+- Open question: Which effects from a parent element are baked into this texture? E.g. things like `filter`, `opacity`, `transform`.
+
+Capturing as a texture avoids the interactivity risks, complexities, and memory impact of fully preserving these parts of Page-A as live DOM.
+
+The root is always captured in this way, and is also clipped to the viewport, as capturing the entire page would take an enormous amount of memory in some cases.
+
+- Open question: Should we have a way to expand this area for particular transitions? For example, transitions that involve vertical movement?
+- Open question: Other elements can also be massive. Do we need a way to limit and control the captured size of those?
+
+This mode works great for the share button and the root, as their transitions can be represented by simple transforms. However, the header changes size without stretching its shadow, and the content of the header moves independently. There's another mode for that:
```

Comment:
I don't think talking about element() is getting into API, it's about implementation. When we talk about representing the element as an image, it helps to clarify that we don't need to define that behaviour from scratch. The modifications we're suggesting are minor and make sense even outside the context of this feature.

Reply:
Let's decide on calling it something with an explanation when the term is introduced. Since we also want to use that term for the replaced element that's displaying this content. I'm comfortable with image since what we cache can be a vector graphic or rasterized pixels and there is precedent for using "image" for both. This is aside from the element() aspect.

And, okay if it seems like this is too early in the doc to say this is element(). As long as when we do mention it, we're indicating that this is the confident option we're advocating for and not an open question.

Reply:
It is not too early to say element(). If we are not specific about a proposal the CSSWG will not know what to review.

That's why when we go to them we need to have a precise thing that is being proposed, which should be:

* element() for getting images of things, with list of edits to element()
* top layer for what it means to be "on top"
* pseudo elements for author styling and UA stylesheet styling

Those are the key things we need to go to them with for alignment.

Reply:
The key thing for the discussion with them is to map everything into CSS spec concepts, to make clear that this is not a scary new thing but builds incrementally on existing things in a layered way.

Reply:
This does not need to be the exposed API, it's the CSS concepts that explain the entire thing. The purpose of spelling out style sheets and element() in the UA style sheet is proof that the layering works.

Reply:
Sounds good.

> element() for getting images of things, with list of edits to element()

Other than the to paint the border or just the contents type of thing, I wanted to mention one other thing to call out about element(). The element() function allows selecting unrendered paint sources (like canvas with display:none). We shouldn't share/render these.

For rendered elements, it lists `an element that is rendered, is not a descendant of a replaced element, and generates a stacking context` as being able to be captured. I'm not sure why the descendant of a replaced element clause is there. It'd be nice to find out. I know things like canvas can have fallback content, but I'm not entirely sure in which situations do we render it

Reply:
I'm going to add a quick reference to `element()` earlier to reassure the reader that this isn't a totally new concept, but I'll also signal that it's not completely the same and the differences are discussed later.